text/0000-delegation.md

```diff
@@ -0,0 +1,612 @@
+>
+> @steven099: It’s used extensively in Objective C / Swift. The thing is that the _delegate_ (n. /ˈdɛlɪɡət/) is the value/type to which you _delegate_ (v. /ˈdɛlɪˌɡeɪt/) functionality. C# has /ˈdɛlɪɡət/s, while I believe this proposal is about /ˈdɛlɪˌɡeɪt/ing.
+
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The biggest non-syntax alternative is only supporting delegation of methods, and not associated types or consts. The authors prefer to support all “trait items” because the whole point is to “make trivial wrapper impls trivial,” even if you’re implementing a trait like Iterator which has an associated type as well as several methods.
+
+### Alternative syntax
+
+```rust
+impl TR for S {
+    delegate to f for *;
+}
+
+impl TR for S {
+    delegate to f for fn foo, fn bar, const MAX, type Item;
+}
+```
+
+The transition to delegate block syntax isn't quite as nice, but it still works:
+
+```rust
+delegate fn foo, fn bar {
+    |&self| self.get_inner(),
+    |&mut self| self.get_inner_mut(),
+    |self| self.into_inner(),
+    |x: Rc<Self>| self.rc_into_inner_rc(),
+} -> {
+    |delegate| Self::from_inner(delegate),
+    |x: Rc<Delegate>| Self::rc_from_inner_rc(x)
+}
+```
+
+The argument for this syntax is "when trait items are explicitly listed rather than globbed, the line quickly becomes long and difficult to read." However, this is solved by the `rustfmt` default of moving `to field_name;` to a new line when delegating individual trait items, e.g.
+
+```rust
+impl TR for S {
+    delegate fn foo, fn bar, fn baz, const MAX, type Item, type Output
+        to really_long_field_name;
+}
+```
+
+### Including `self` in delegate statements
+
+```rust
+impl TR for S {
+    delegate * to self.f;
+}
+
+impl TR for S {
+    delegate to self.f for *;
+}
+```
+
+This makes it more obvious the target is a struct field and is certainly more clear for tuple structs. This would be a good choice if we want to allow custom implementer expressions without a delegate block, when there is no variance due to the type of self.
+
+`self.` is unnecessary and could lead newcomers to believe they can write arbitrary Rust code there. However, this risk also applies to allowing getter methods in the basic delegation statement.
+
+This unresolved question is relevant to the decision:
+-   Are there any cases of [Custom Self Types](https://github.com/rust-lang/rfcs/pull/2362) where self needs to be manually dereferenced, e.g.
+    ```rust
+    impl TR for S {
+        fn foo(self: Arc<Self>) -> u32 {
+            *self.f.foo()
+        }
+    }
+    ```
+
+
+### Omitting the `impl` block
+
+The `impl` block is a lot of extra noise when simply delegating all trait items to a struct field. There can be a lot of generic type parameters which are copy-pasted from the member field's trait impl.
+
+```rust
+delegate TR::* to S::f;
```

Comment:
I would suggest going even a step further and allowed also `delegate TR to S::f` and `delegate TR1, TR2 to S::f`, because I suspect it will be pretty common to delegate several traits, in their entirety, to a field. Traits that have just one required method are quite common and many of them have other related traits, so you are likely to be delegating several traits at once.